File: tests/reference/arc.py
"""Independent slow reference for E3 procedural ARC-like grids (spec 7.1, 10, 15.5).

This is the source of truth for dihedral / color-perm / augment / tokenize /
family generate / sample_family_specs / diversity_stats / ProceduralCorpus.sample.
Production ``synth`` must match these results on the same inputs. Production must
never import ``tests/``. Must match ``synth/tests/common/arc_ref.rs``.

Pinned rules
------------
Dihedral (D4), index 0..8
    0 identity
    1 rot90 clockwise. ``[[1,2,3],[4,5,6]]`` -> ``[[4,1],[5,2],[6,3]]``.
    2 rot180
    3 rot270 clockwise
    4 flip left-right (horizontal)
    5 flip up-down (vertical)
    6 transpose (main diagonal)
    7 anti-transpose (anti-diagonal): transpose then rot180.
    rot90 / rot270 / transpose / anti-transpose swap height and width.
    After any dihedral the grid still passes Grid validation.

Color permutation
    ``perm[c]`` is the new color of ``c``. Length 10, each of 0..9 exactly once.
    Color 0 is not special. Bad perm -> SynthError matching Rust BadPermutation.

Augmentation
    Dihedral first, then color perm. Same transform on every grid in the pair/task.
    ``family_id`` and ``seed`` unchanged.

tokenize_grid
    ``tok.encode_grid(grid.flatten())`` as a list. Tokenizer exceptions propagate.

Family params (exact keys; anything else is ``unknown family``)
    translate: dx, dy, bg. Non-bg cells move by (dx, dy) with wrap-around
        (torus). Output size equals input. bg in 0..10; dx, dy any int.
        Positive dx is right, positive dy is down.
    recolor: src, dst. Map src to dst. src != dst, both in 0..10.
    crop: bg. Bounding box of non-bg cells. All-bg input -> 1x1 ``[[bg]]``.
    tile: nx, ny. Repeat nx times horizontally and ny vertically. nx, ny in 1..30.
    gravity: dir (0 down, 1 up, 2 left, 3 right), bg. Non-bg cells pack toward
        that edge in scan order; bg fills the rest.
    mirror: dihedral (0..8). Output is apply_dihedral(input, that index).
    scale: factor 2 or 3. Cell replication. Output size = input * factor.
    border: color in 0..10, width 1 or 2. Output grows by 2*width per axis;
        interior is the input, frame is ``color``.

generate(seed) / generate_task(spec, seed, n_train)
    SplitMix64(seed). n_train == 0 -> empty train. Else sample exactly
    n_train + 1 distinct pairs (retry on collision, cap 10000). Train is the
    first n_train; test is the last. Test is not identical to any train pair.
    Family.generate uses DEFAULT_N_TRAIN=3. Task.seed is the generate seed.
    Input size: uniform 1..=min(8, max that keeps the transformed grid <= 30).
    Each cell uniform in 0..10. Same (kind, params, seed, n_train) matches Rust.

sample_family_specs(n, seed)
    n == 0 -> []. Shuffle the parameter grid with SplitMix64(seed) Fisher-Yates
    and take n. n larger than the grid -> SynthError message listed in
    TOO_MANY_SPECS. Cover all 8 kinds when n equals the grid length.

diversity_stats
    unique_tasks hashes train+test grid cells only (not family_id / seed).
    Empty tasks: all zeros, collision_rate 0.0.

ProceduralCorpus.sample(n)
    n == 0 -> empty tasks (even if specs is empty).
    Empty specs -> unknown family.
    n_train <= 0 -> empty train.
    Else task i uses specs[i % len] and seed+i with corpus n_train.
"""

from __future__ import annotations

from collections.abc import Sequence

from synth import (
    DEFAULT_N_TRAIN,
    MAX_GRID_SIZE,
    N_COLORS,
    N_DIHEDRAL,
    Dihedral,
    DiversityStats,
    FamilyKind,
    FamilySpec,
    Grid,
    Pair,
    SynthError,
    Task,
)

TOO_MANY_SPECS = "requested more family specs than the parameter grid"
DISTINCT_PAIR_FAIL = "could not sample distinct pairs"
MAX_SAMPLE_DIM = 8
MAX_PAIR_ATTEMPTS = 10000
_MASK64 = (1 << 64) - 1
_SPLITMIX_GOLDEN = 0x9E3779B97F4A7C15
_SPLITMIX_M1 = 0xBF58476D1CE4E5B9
_SPLITMIX_M2 = 0x94D049BB133111EB


class SplitMix64:
    """Portable PRNG. Must match synth/tests/common/arc_ref.rs."""

    __slots__ = ("state",)

    def __init__(self, seed: int) -> None:
        self.state = seed & _MASK64

    def next_u64(self) -> int:
        self.state = (self.state + _SPLITMIX_GOLDEN) & _MASK64
        z = self.state
        z = ((z ^ (z >> 30)) * _SPLITMIX_M1) & _MASK64
        z = ((z ^ (z >> 27)) * _SPLITMIX_M2) & _MASK64
        return z ^ (z >> 31)

    def next_bounded(self, n: int) -> int:
        return self.next_u64() % n


class RefFamily:
    """Reference family. ``generate`` always uses DEFAULT_N_TRAIN."""

    def __init__(self, spec: FamilySpec) -> None:
        self._spec = spec
        self._id = spec.family_id()

    def id(self) -> str:
        return self._id

    def kind(self) -> FamilyKind:
        return self._spec.kind

    def generate(self, seed: int) -> Task:
        return generate_task(self._spec, seed, DEFAULT_N_TRAIN)


def apply_dihedral(grid: Grid, dihedral: Dihedral) -> Grid:
    src = grid.cells
    h = len(src)
    w = len(src[0])
    k = dihedral.index()
    if k == 0:
        oh, ow = h, w

        def cell(r: int, c: int) -> int:
            return src[r][c]
    elif k == 1:
        oh, ow = w, h

        def cell(r: int, c: int) -> int:
            return src[h - 1 - c][r]
    elif k == 2:
        oh, ow = h, w

        def cell(r: int, c: int) -> int:
            return src[h - 1 - r][w - 1 - c]
    elif k == 3:
        oh, ow = w, h

        def cell(r: int, c: int) -> int:
            return src[c][w - 1 - r]
    elif k == 4:
        oh, ow = h, w

        def cell(r: int, c: int) -> int:
            return src[r][w - 1 - c]
    elif k == 5:
        oh, ow = h, w

        def cell(r: int, c: int) -> int:
            return src[h - 1 - r][c]
    elif k == 6:
        oh, ow = w, h

        def cell(r: int, c: int) -> int:
            return src[c][r]
    elif k == 7:
        oh, ow = w, h

        def cell(r: int, c: int) -> int:
            return src[h - 1 - c][w - 1 - r]
    else:
        raise SynthError(f"dihedral index {k} out of range")
    return Grid(cells=[[cell(r, c) for c in range(ow)] for r in range(oh)])


def apply_color_perm(grid: Grid, perm: Sequence[int]) -> Grid:
    p = [int(x) for x in perm]
    if len(p) != N_COLORS:
        raise SynthError("color permutation is not a permutation of 0..10")
    seen = [False] * N_COLORS
    for v in p:
        if v < 0 or v >= N_COLORS or seen[v]:
            raise SynthError("color permutation is not a permutation of 0..10")
        seen[v] = True
    return Grid(cells=[[p[c] for c in row] for row in grid.cells])


def augment_pair(pair: Pair, dihedral: Dihedral, perm: Sequence[int]) -> Pair:
    return Pair(
        input=apply_color_perm(apply_dihedral(pair.input, dihedral), perm),
        output=apply_color_perm(apply_dihedral(pair.output, dihedral), perm),
    )


def augment_task(task: Task, dihedral: Dihedral, perm: Sequence[int]) -> Task:
    return Task(
        family_id=task.family_id,
        seed=task.seed,
        train=[augment_pair(p, dihedral, perm) for p in task.train],
        test=augment_pair(task.test, dihedral, perm),
    )


def tokenize_grid(tok: object, grid: Grid) -> list[int]:
    return list(tok.encode_grid(grid.flatten()))  # type: ignore[attr-defined]


def _require(spec: FamilySpec, keys: set[str]) -> dict[str, int]:
    d = dict(spec.params)
    if set(d) != keys:
        raise SynthError("unknown family")
    return d


def _color(v: int) -> int:
    if v < 0 or v >= N_COLORS:
        raise SynthError("unknown family")
    return v


def validate_spec(spec: FamilySpec) -> dict[str, int]:
    """Return params if ``spec`` is a known family, else ``unknown family``."""
    kind = spec.kind
    if kind is FamilyKind.TRANSLATE:
        d = _require(spec, {"dx", "dy", "bg"})
        _color(d["bg"])
        return d
    if kind is FamilyKind.RECOLOR:
        d = _require(spec, {"src", "dst"})
        src, dst = _color(d["src"]), _color(d["dst"])
        if src == dst:
            raise SynthError("unknown family")
        return d
    if kind is FamilyKind.CROP:
        d = _require(spec, {"bg"})
        _color(d["bg"])
        return d
    if kind is FamilyKind.TILE:
        d = _require(spec, {"nx", "ny"})
        if d["nx"] < 1 or d["ny"] < 1 or d["nx"] > MAX_GRID_SIZE or d["ny"] > MAX_GRID_SIZE:
            raise SynthError("unknown family")
        return d
    if kind is FamilyKind.GRAVITY:
        d = _require(spec, {"dir", "bg"})
        if d["dir"] < 0 or d["dir"] > 3:
            raise SynthError("unknown family")
        _color(d["bg"])
        return d
    if kind is FamilyKind.MIRROR:
        d = _require(spec, {"dihedral"})
        if d["dihedral"] < 0 or d["dihedral"] >= N_DIHEDRAL:
            raise SynthError("unknown family")
        return d
    if kind is FamilyKind.SCALE:
        d = _require(spec, {"factor"})
        if d["factor"] not in (2, 3):
            raise SynthError("unknown family")
        return d
    if kind is FamilyKind.BORDER:
        d = _require(spec, {"color", "width"})
        _color(d["color"])
        if d["width"] not in (1, 2):
            raise SynthError("unknown family")
        return d
    raise SynthError("unknown family")


def _wrap(i: int, n: int) -> int:
    r = i % n
    if r < 0:
        r += n
    return r


def apply_family(grid: Grid, spec: FamilySpec) -> Grid:
    d = validate_spec(spec)
    kind = spec.kind
    if kind is FamilyKind.TRANSLATE:
        return _translate(grid, d["dx"], d["dy"], d["bg"])
    if kind is FamilyKind.RECOLOR:
        return _recolor(grid, d["src"], d["dst"])
    if kind is FamilyKind.CROP:
        return _crop(grid, d["bg"])
    if kind is FamilyKind.TILE:
        return _tile(grid, d["nx"], d["ny"])
    if kind is FamilyKind.GRAVITY:
        return _gravity(grid, d["dir"], d["bg"])
    if kind is FamilyKind.MIRROR:
        return apply_dihedral(grid, Dihedral.from_index(d["dihedral"]))
    if kind is FamilyKind.SCALE:
        return _scale(grid, d["factor"])
    return _border(grid, d["color"], d["width"])


def _translate(grid: Grid, dx: int, dy: int, bg: int) -> Grid:
    h, w = grid.rows(), grid.cols()
    out = [[bg for _ in range(w)] for _ in range(h)]
    for r, row in enumerate(grid.cells):
        for c, val in enumerate(row):
            if val != bg:
                out[_wrap(r + dy, h)][_wrap(c + dx, w)] = val
    return Grid(cells=out)


def _recolor(grid: Grid, src: int, dst: int) -> Grid:
    return Grid(cells=[[dst if c == src else c for c in row] for row in grid.cells])


def _crop(grid: Grid, bg: int) -> Grid:
    coords = [(r, c) for r, row in enumerate(grid.cells) for c, val in enumerate(row) if val != bg]
    if not coords:
        return Grid(cells=[[bg]])
    min_r = min(r for r, _ in coords)
    max_r = max(r for r, _ in coords)
    min_c = min(c for _, c in coords)
    max_c = max(c for _, c in coords)
    return Grid(cells=[row[min_c : max_c + 1] for row in grid.cells[min_r : max_r + 1]])


def _tile(grid: Grid, nx: int, ny: int) -> Grid:
    h = grid.rows()
    out: list[list[int]] = []
    for _tr in range(ny):
        for r in range(h):
            row: list[int] = []
            for _tc in range(nx):
                row.extend(grid.cells[r])
            out.append(row)
    return Grid(cells=out)


def _gravity(grid: Grid, direction: int, bg: int) -> Grid:
    h, w = grid.rows(), grid.cols()
    out = [[bg for _ in range(w)] for _ in range(h)]
    if direction == 0:
        for c in range(w):
            objs = [grid.cells[r][c] for r in range(h) if grid.cells[r][c] != bg]
            start = h - len(objs)
            for i, val in enumerate(objs):
                out[start + i][c] = val
    elif direction == 1:
        for c in range(w):
            objs = [grid.cells[r][c] for r in range(h) if grid.cells[r][c] != bg]
            for i, val in enumerate(objs):
                out[i][c] = val
    elif direction == 2:
        for r in range(h):
            objs = [val for val in grid.cells[r] if val != bg]
            for i, val in enumerate(objs):
                out[r][i] = val
    else:
        for r in range(h):
            objs = [val for val in grid.cells[r] if val != bg]
            start = w - len(objs)
            for i, val in enumerate(objs):
                out[r][start + i] = val
    return Grid(cells=out)


def _scale(grid: Grid, factor: int) -> Grid:
    out = []
    for row in grid.cells:
        scaled = [c for c in row for _ in range(factor)]
        for _ in range(factor):
            out.append(list(scaled))
    return Grid(cells=out)


def _border(grid: Grid, color: int, width: int) -> Grid:
    h, w = grid.rows(), grid.cols()
    oh, ow = h + 2 * width, w + 2 * width
    out = [[color for _ in range(ow)] for _ in range(oh)]
    for r, row in enumerate(grid.cells):
        for c, val in enumerate(row):
            out[r + width][c + width] = val
    return Grid(cells=out)


def family_from_spec(spec: FamilySpec) -> RefFamily:
    validate_spec(spec)
    return RefFamily(spec)


def _max_in_hw(spec: FamilySpec) -> tuple[int, int]:
    d = dict(spec.params)
    kind = spec.kind
    if kind is FamilyKind.TILE:
        return MAX_GRID_SIZE // d["ny"], MAX_GRID_SIZE // d["nx"]
    if kind is FamilyKind.SCALE:
        f = d["factor"]
        return MAX_GRID_SIZE // f, MAX_GRID_SIZE // f
    if kind is FamilyKind.BORDER:
        w = d["width"]
        return MAX_GRID_SIZE - 2 * w, MAX_GRID_SIZE - 2 * w
    return MAX_GRID_SIZE, MAX_GRID_SIZE


def _cells_key(grid: Grid) -> tuple[tuple[int, ...], ...]:
    return tuple(tuple(row) for row in grid.cells)


def _sample_pair(rng: SplitMix64, spec: FamilySpec) -> Pair:
    max_h, max_w = _max_in_hw(spec)
    max_h = min(MAX_SAMPLE_DIM, max_h)
    max_w = min(MAX_SAMPLE_DIM, max_w)
    h = 1 + rng.next_bounded(max_h)
    w = 1 + rng.next_bounded(max_w)
    cells = [[rng.next_bounded(N_COLORS) for _ in range(w)] for _ in range(h)]
    inp = Grid(cells=cells)
    return Pair(input=inp, output=apply_family(inp, spec))


def generate_task(spec: FamilySpec, seed: int, n_train: int) -> Task:
    validate_spec(spec)
    if n_train <= 0:
        raise SynthError("empty train")
    rng = SplitMix64(seed)
    need = n_train + 1
    pairs: list[Pair] = []
    seen: set[tuple] = set()
    attempts = 0
    while len(pairs) < need:
        attempts += 1
        if attempts > MAX_PAIR_ATTEMPTS:
            raise SynthError(DISTINCT_PAIR_FAIL)
        pair = _sample_pair(rng, spec)
        key = (_cells_key(pair.input), _cells_key(pair.output))
        if key in seen:
            continue
        seen.add(key)
        pairs.append(pair)
    return Task(
        family_id=spec.family_id(),
        seed=seed & _MASK64,
        train=pairs[:-1],
        test=pairs[-1],
    )


def parameter_grid() -> list[FamilySpec]:
    specs: list[FamilySpec] = []
    for bg in range(N_COLORS):
        for dx in range(-3, 4):
            for dy in range(-3, 4):
                specs.append(FamilySpec.make(FamilyKind.TRANSLATE, {"dx": dx, "dy": dy, "bg": bg}))
    for src in range(N_COLORS):
        for dst in range(N_COLORS):
            if src != dst:
                specs.append(FamilySpec.make(FamilyKind.RECOLOR, {"src": src, "dst": dst}))
    for bg in range(N_COLORS):
        specs.append(FamilySpec.make(FamilyKind.CROP, {"bg": bg}))
    for nx in range(1, 5):
        for ny in range(1, 5):
            specs.append(FamilySpec.make(FamilyKind.TILE, {"nx": nx, "ny": ny}))
    for bg in range(N_COLORS):
        for direction in range(4):
            specs.append(FamilySpec.make(FamilyKind.GRAVITY, {"dir": direction, "bg": bg}))
    for d in range(N_DIHEDRAL):
        specs.append(FamilySpec.make(FamilyKind.MIRROR, {"dihedral": d}))
    for factor in (2, 3):
        specs.append(FamilySpec.make(FamilyKind.SCALE, {"factor": factor}))
    for color in range(N_COLORS):
        for width in (1, 2):
            specs.append(FamilySpec.make(FamilyKind.BORDER, {"color": color, "width": width}))
    return specs


def sample_family_specs(n: int, seed: int) -> list[FamilySpec]:
    if n == 0:
        return []
    specs = parameter_grid()
    if n < 0 or n > len(specs):
        raise SynthError(TOO_MANY_SPECS)
    rng = SplitMix64(seed)
    for i in range(len(specs)):
        j = i + rng.next_bounded(len(specs) - i)
        specs[i], specs[j] = specs[j], specs[i]
    return specs[:n]


def diversity_stats(tasks: Sequence[Task]) -> DiversityStats:
    n_tasks = len(tasks)
    families: set[str] = set()
    test_in: set[tuple] = set()
    test_out: set[tuple] = set()
    task_keys: set[tuple] = set()
    hist = [0] * N_COLORS
    shapes: set[tuple[int, int]] = set()

    def acc(grid: Grid) -> None:
        shapes.add((grid.rows(), grid.cols()))
        for row in grid.cells:
            for c in row:
                hist[c] += 1

    for task in tasks:
        families.add(task.family_id)
        test_in.add(_cells_key(task.test.input))
        test_out.add(_cells_key(task.test.output))
        train_key = tuple((_cells_key(p.input), _cells_key(p.output)) for p in task.train)
        task_keys.add((train_key, _cells_key(task.test.input), _cells_key(task.test.output)))
        for pair in task.train:
            acc(pair.input)
            acc(pair.output)
        acc(task.test.input)
        acc(task.test.output)

    unique_tasks = len(task_keys)
    collision = 0.0 if n_tasks == 0 else 1.0 - unique_tasks / n_tasks
    return DiversityStats(
        n_tasks=n_tasks,
        n_families=len(families),
        unique_test_inputs=len(test_in),
        unique_test_outputs=len(test_out),
        unique_tasks=unique_tasks,
        color_histogram=tuple(hist),
        unique_shapes=len(shapes),
        collision_rate=collision,
    )


def sample_corpus(specs: Sequence[FamilySpec], seed: int, n_train: int, n: int) -> list[Task]:
    if n == 0:
        raise SynthError("empty tasks")
    if not specs:
        raise SynthError("unknown family")
    if n_train <= 0:
        raise SynthError("empty train")
    out: list[Task] = []
    for i in range(n):
        spec = specs[i % len(specs)]
        out.append(generate_task(spec, seed + i, n_train))
    return out

File: synth/src/arc.rs
//! Procedural ARC-like grids (spec 7.1, 10, 15.5 E3).
//!
//! re-arc-style generators: parameterized task families produce input/output
//! grid pairs. Cells are palette indices `0..N_COLORS`. One token per cell
//! via F6 `encode_grid` on the row-major flatten. Training-time augmentation
//! is the 8 dihedral transforms times color permutations. The gate is
//! [`diversity_stats`].
//!
//! Grids are structurally the D2 `Grid` (`cells: Vec<Vec<u8>>`) but this
//! crate does not depend on `prometheus-verifiers`. Tokenization wraps F6
//! behind [`GridTokenizer`].

use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};

use crate::{Error, Result};

/// ARC-AGI cell colors 0-9. Matches F6 `ARC_N_COLORS`.
pub const N_COLORS: u8 = 10;
/// Inclusive minimum rows and cols.
pub const MIN_GRID_SIZE: usize = 1;
/// Inclusive maximum rows and cols. ARC-AGI official max.
pub const MAX_GRID_SIZE: usize = 30;
/// Order of the dihedral group D4.
pub const N_DIHEDRAL: u8 = 8;
/// Default demonstration pairs per task.
pub const DEFAULT_N_TRAIN: u32 = 3;

/// One ARC-style grid. Cell values are palette indices in `0..N_COLORS`.
/// Rectangular, rows and cols in `MIN_GRID_SIZE..=MAX_GRID_SIZE`.
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct Grid {
    pub cells: Vec<Vec<u8>>,
}

impl Grid {
    /// Validate and wrap. Empty, jagged, out-of-range size, or a color
    /// outside `0..N_COLORS` is an error.
    pub fn try_new(cells: Vec<Vec<u8>>) -> Result<Self> {
        if cells.is_empty() || cells[0].is_empty() {
            return Err(Error::EmptyGrid);
        }
        let rows = cells.len();
        let cols = cells[0].len();
        if !(MIN_GRID_SIZE..=MAX_GRID_SIZE).contains(&rows)
            || !(MIN_GRID_SIZE..=MAX_GRID_SIZE).contains(&cols)
        {
            return Err(Error::GridSize { rows, cols });
        }
        for row in &cells {
            if row.len() != cols {
                return Err(Error::JaggedGrid);
            }
            for &c in row {
                if c >= N_COLORS {
                    return Err(Error::BadColor(c));
                }
            }
        }
        Ok(Self { cells })
    }

    pub fn rows(&self) -> usize {
        self.cells.len()
    }

    pub fn cols(&self) -> usize {
        self.cells.first().map(|r| r.len()).unwrap_or(0)
    }

    /// Row-major flatten. This is the F6 `encode_grid` input.
    pub fn flatten(&self) -> Vec<u8> {
        self.cells.iter().flatten().copied().collect()
    }
}

/// One input/output pair.
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct Pair {
    pub input: Grid,
    pub output: Grid,
}

/// One generated task. `train` are demonstrations; `test` is held out.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Task {
    pub family_id: String,
    pub seed: u64,
    pub train: Vec<Pair>,
    pub test: Pair,
}

/// D4 action. JSON names are snake_case values. Indices 0..8:
/// 0 identity, 1 rot90 CW, 2 rot180, 3 rot270 CW, 4 flip left-right,
/// 5 flip up-down, 6 transpose (main diagonal), 7 anti-transpose.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Dihedral {
    Identity,
    Rot90,
    Rot180,
    Rot270,
    FlipH,
    FlipV,
    Transpose,
    AntiTranspose,
}

impl Dihedral {
    pub const ALL: [Dihedral; 8] = [
        Dihedral::Identity,
        Dihedral::Rot90,
        Dihedral::Rot180,
        Dihedral::Rot270,
        Dihedral::FlipH,
        Dihedral::FlipV,
        Dihedral::Transpose,
        Dihedral::AntiTranspose,
    ];

    pub fn as_str(self) -> &'static str {
        match self {
            Dihedral::Identity => "identity",
            Dihedral::Rot90 => "rot90",
            Dihedral::Rot180 => "rot180",
            Dihedral::Rot270 => "rot270",
            Dihedral::FlipH => "flip_h",
            Dihedral::FlipV => "flip_v",
            Dihedral::Transpose => "transpose",
            Dihedral::AntiTranspose => "anti_transpose",
        }
    }

    pub fn index(self) -> u8 {
        match self {
            Dihedral::Identity => 0,
            Dihedral::Rot90 => 1,
            Dihedral::Rot180 => 2,
            Dihedral::Rot270 => 3,
            Dihedral::FlipH => 4,
            Dihedral::FlipV => 5,
            Dihedral::Transpose => 6,
            Dihedral::AntiTranspose => 7,
        }
    }

    pub fn from_index(k: u8) -> Result<Self> {
        Dihedral::ALL
            .get(k as usize)
            .copied()
            .ok_or(Error::BadDihedral(k))
    }

    pub fn parse(s: &str) -> Result<Self> {
        match s {
            "identity" => Ok(Dihedral::Identity),
            "rot90" => Ok(Dihedral::Rot90),
            "rot180" => Ok(Dihedral::Rot180),
            "rot270" => Ok(Dihedral::Rot270),
            "flip_h" => Ok(Dihedral::FlipH),
            "flip_v" => Ok(Dihedral::FlipV),
            "transpose" => Ok(Dihedral::Transpose),
            "anti_transpose" => Ok(Dihedral::AntiTranspose),
            other => Err(Error::Message(format!("unknown dihedral {other}"))),
        }
    }
}

/// Primitive generator kind. Distinct `(kind, params)` pairs are distinct
/// families. Spec 7.1 wants millions of families via parameterization, not
/// 400 hand-written ARC clones.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum FamilyKind {
    /// Move the non-background object by `(dx, dy)`. Params: `dx`, `dy`, `bg`.
    Translate,
    /// Map color `src` to `dst`. Params: `src`, `dst`.
    Recolor,
    /// Crop to the bounding box of non-`bg` cells. Params: `bg`.
    Crop,
    /// Repeat the input `nx` by `ny` times. Params: `nx`, `ny`.
    Tile,
    /// Drop non-`bg` cells along `dir` (0 down, 1 up, 2 left, 3 right).
    /// Params: `dir`, `bg`.
    Gravity,
    /// Output is a dihedral of the input. Params: `dihedral` (0..8).
    Mirror,
    /// Nearest-neighbor upscale. Params: `factor` (2 or 3).
    Scale,
    /// Draw a border. Params: `color`, `width` (1 or 2).
    Border,
}

impl FamilyKind {
    pub const ALL: [FamilyKind; 8] = [
        FamilyKind::Translate,
        FamilyKind::Recolor,
        FamilyKind::Crop,
        FamilyKind::Tile,
        FamilyKind::Gravity,
        FamilyKind::Mirror,
        FamilyKind::Scale,
        FamilyKind::Border,
    ];

    pub fn as_str(self) -> &'static str {
        match self {
            FamilyKind::Translate => "translate",
            FamilyKind::Recolor => "recolor",
            FamilyKind::Crop => "crop",
            FamilyKind::Tile => "tile",
            FamilyKind::Gravity => "gravity",
            FamilyKind::Mirror => "mirror",
            FamilyKind::Scale => "scale",
            FamilyKind::Border => "border",
        }
    }

    pub fn parse(s: &str) -> Result<Self> {
        match s {
            "translate" => Ok(FamilyKind::Translate),
            "recolor" => Ok(FamilyKind::Recolor),
            "crop" => Ok(FamilyKind::Crop),
            "tile" => Ok(FamilyKind::Tile),
            "gravity" => Ok(FamilyKind::Gravity),
            "mirror" => Ok(FamilyKind::Mirror),
            "scale" => Ok(FamilyKind::Scale),
            "border" => Ok(FamilyKind::Border),
            other => Err(Error::Message(format!("unknown family {other}"))),
        }
    }
}

/// Parameterized family. `params` keys are documented on [`FamilyKind`].
/// `family_id` is `"{kind}:{k}={v},..."` with keys sorted.
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct FamilySpec {
    pub kind: FamilyKind,
    pub params: BTreeMap<String, i32>,
}

impl FamilySpec {
    pub fn new(kind: FamilyKind, params: BTreeMap<String, i32>) -> Self {
        Self { kind, params }
    }

    pub fn family_id(&self) -> String {
        let mut parts: Vec<String> = self
            .params
            .iter()
            .map(|(k, v)| format!("{k}={v}"))
            .collect();
        parts.sort();
        if parts.is_empty() {
            self.kind.as_str().to_string()
        } else {
            format!("{}:{}", self.kind.as_str(), parts.join(","))
        }
    }
}

/// Generator for one family. `generate` is deterministic in `seed`.
pub trait Family {
    fn id(&self) -> &str;
    fn kind(&self) -> FamilyKind;
    fn generate(&self, seed: u64) -> Result<Task>;
}

/// F6 `encode_grid`. Implementors wrap `tokenizer.Tokenizer.encode_grid`.
pub trait GridTokenizer {
    fn encode_grid(&self, cells: &[u8]) -> std::result::Result<Vec<u32>, String>;
}

/// Gate numbers. `collision_rate` is `1 - unique_tasks / n_tasks`, or 0.0
/// when `n_tasks == 0`. Hashes are in-process identity of the grid cells,
/// not a stable digest.
#[derive(Debug, Clone, PartialEq)]
pub struct DiversityStats {
    pub n_tasks: u64,
    pub n_families: u64,
    pub unique_test_inputs: u64,
    pub unique_test_outputs: u64,
    pub unique_tasks: u64,
    pub color_histogram: [u64; 10],
    pub unique_shapes: u64,
    pub collision_rate: f64,
}

/// Apply a D4 action. `rot90` is clockwise. Output size swaps on rot90,
/// rot270, transpose, anti-transpose.
pub fn apply_dihedral(grid: &Grid, dihedral: Dihedral) -> Result<Grid> {
    let _ = (grid, dihedral);
    unimplemented!("E3 apply_dihedral")
}

/// `perm[c]` is the new color of `c`. Must be a permutation of `0..N_COLORS`.
pub fn apply_color_perm(grid: &Grid, perm: &[u8]) -> Result<Grid> {
    let _ = (grid, perm);
    unimplemented!("E3 apply_color_perm")
}

/// Same dihedral and color perm on every grid in the pair.
pub fn augment_pair(pair: &Pair, dihedral: Dihedral, perm: &[u8]) -> Result<Pair> {
    let _ = (pair, dihedral, perm);
    unimplemented!("E3 augment_pair")
}

/// Same dihedral and color perm on every grid in the task. `family_id` and
/// `seed` are unchanged. Augmentation is a training-time view, not a new
/// family.
pub fn augment_task(task: &Task, dihedral: Dihedral, perm: &[u8]) -> Result<Task> {
    let _ = (task, dihedral, perm);
    unimplemented!("E3 augment_task")
}

/// F6 encode of `grid.flatten()`.
pub fn tokenize_grid(tok: &dyn GridTokenizer, grid: &Grid) -> Result<Vec<u32>> {
    let _ = (tok, grid);
    unimplemented!("E3 tokenize_grid")
}

/// Construct a family from a spec. Unknown or invalid params are
/// [`Error::UnknownFamily`].
pub fn family_from_spec(spec: FamilySpec) -> Result<Box<dyn Family>> {
    let _ = spec;
    unimplemented!("E3 family_from_spec")
}

/// `n` distinct specs from the parameter grid, deterministic in `seed`.
pub fn sample_family_specs(n: u32, seed: u64) -> Result<Vec<FamilySpec>> {
    let _ = (n, seed);
    unimplemented!("E3 sample_family_specs")
}

/// Gate: uniqueness and color/shape coverage over `tasks`. Empty is zeros
/// and `collision_rate == 0.0`.
pub fn diversity_stats(tasks: &[Task]) -> DiversityStats {
    let _ = tasks;
    unimplemented!("E3 diversity_stats")
}

/// A bag of families to sample from.
pub struct ProceduralCorpus {
    pub specs: Vec<FamilySpec>,
    pub seed: u64,
    pub n_train: u32,
}

impl ProceduralCorpus {
    pub fn new(specs: Vec<FamilySpec>, seed: u64) -> Self {
        Self {
            specs,
            seed,
            n_train: DEFAULT_N_TRAIN,
        }
    }

    /// `n` tasks, cycling specs, seeds `seed + i`. Empty `specs` is
    /// [`Error::UnknownFamily`]. `n == 0` is [`Error::EmptyTasks`].
    pub fn sample(&self, n: u32) -> Result<Vec<Task>> {
        let _ = n;
        unimplemented!("E3 ProceduralCorpus::sample")
    }
}
